src/execution_tree/primitives/gradient_operation.cpp

```diff
@@ -0,0 +1,265 @@
+// Copyright (c) 2018 Bibek Wagle
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <phylanx/config.hpp>
+#include <phylanx/execution_tree/primitives/gradient_operation.hpp>
+#include <phylanx/ir/node_data.hpp>
+
+#include <hpx/include/lcos.hpp>
+#include <hpx/include/naming.hpp>
+#include <hpx/include/util.hpp>
+#include <hpx/throw_exception.hpp>
+
+#include <cstddef>
+#include <cstdint>
+#include <cmath>
+#include <memory>
+#include <numeric>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <blaze/Math.h>
+
+///////////////////////////////////////////////////////////////////////////////
+namespace phylanx { namespace execution_tree { namespace primitives
+{
+    ///////////////////////////////////////////////////////////////////////////
+    primitive create_gradient_operation(hpx::id_type const& locality,
+        std::vector<primitive_argument_type>&& operands,
+        std::string const& name, std::string const& codename)
+    {
+        static std::string type("gradient");
+        return create_primitive_component(
+            locality, type, std::move(operands), name, codename);
+    }
+
+    match_pattern_type const gradient_operation::match_data =
+    {
+        hpx::util::make_tuple("gradient",
+            std::vector<std::string>{"gradient(__1)"},
```

Comment:
This will accept any number of arguments. It might be better to use `gradient(_1, _2)` if you need exactly two arguments.

Reply:
I wanted the gradient primitive to take either one or two arguments, the second one being the optional axis. In order to make sure that I get either one or two arguments, I throw an exception if the size of the operands is greater than 2 inside the eval() function.

Reply:
In case you might want to write:
```
match_pattern_type const gradient_operation::match_data = 
{ 
    hpx::util::make_tuple("gradient", 
        std::vector<std::string>{"gradient(_1, _2)", "gradient(_1)"}, 
        &create_gradient_operation, 
        &create_primitive<gradient_operation>) 
}; 
```